File: backend/src/main/scala/domains/user/http/api/ListUserSuggestions.scala
package domains.user.http.api

import cats.effect.IO
import domains.auth.http.AuthenticatedApi
import domains.auth.objects.AuthUser
import domains.user.http.UserApiSupport
import domains.user.http.codec.UserHttpCodecs.given
import domains.user.objects.UserIdentity
import domains.user.objects.request.UserSearchQuery
import domains.user.table.user.UserTable
import io.circe.Encoder
import org.http4s.{Method, Request, Status}
import shared.http.{ApiPath, HttpApiError, PathParams}

import java.sql.Connection

object ListUserSuggestions extends AuthenticatedApi[UserSearchQuery, List[UserIdentity]]:

  override val method: Method = Method.GET
  override val path: ApiPath = ApiPath("/api/users/suggestions")
  override val successStatus: Status = Status.Ok
  override protected val outputEncoder: Encoder[List[UserIdentity]] = summon[Encoder[List[UserIdentity]]]

  override def decode(request: Request[IO], pathParams: PathParams): IO[UserSearchQuery] =
    val _ = pathParams
    HttpApiError.fromEitherBadRequest(UserSearchQuery.parse(request.uri.query.params.get("q").getOrElse("")))

  override def plan(connection: Connection, actor: AuthUser, query: UserSearchQuery): IO[List[UserIdentity]] =
    val _ = actor
    if query.value.length < UserApiSupport.minSuggestionQueryLength then IO.pure(Nil)
    else UserTable.listSuggestions(connection, query)

File: frontend/src/apis/user/ListUserSuggestions.ts
import type { APIMessage } from '@/system/api/api-message'
import type { UserIdentity } from '@/objects/user/UserIdentity'
import type { UserSearchQuery } from '@/objects/user/request/UserSearchQuery'

/** 查询用户搜索建议；输入已解析搜索词，输出公开用户身份列表。 */
export class ListUserSuggestions implements APIMessage<UserIdentity[]> {
  declare readonly responseType?: UserIdentity[]
  readonly method = 'GET'
  readonly apiPath: string

  constructor(query: UserSearchQuery) {
    const params = new URLSearchParams()
    params.set('q', query)
    this.apiPath = `users/suggestions?${params.toString()}`
  }

  body(): undefined {
    return undefined
  }
}
